File: helper/src/main/java/com/memoeslink/helper/SharedPreferencesHelper.java
package com.memoeslink.helper;

import android.content.Context;
import android.content.ContextWrapper;
import android.content.SharedPreferences;

import androidx.preference.PreferenceManager;

import java.util.Collections;
import java.util.Map;
import java.util.Set;

public class SharedPreferencesHelper extends ContextWrapper {
    public static final String PREFERENCES = "prefs";
    private static final Boolean DEFAULT_BOOLEAN = false;
    private static final Float DEFAULT_FLOAT = 0F;
    private static final Integer DEFAULT_INTEGER = 0;
    private static final Long DEFAULT_LONG = 0L;
    private static final String DEFAULT_STRING = "";
    private static final Set<String> DEFAULT_STRING_SET = Collections.emptySet();
    private final SharedPreferences preferences;
    private final SharedPreferences.Editor editor;

    public SharedPreferencesHelper(Context context) {
        super(context);
        preferences = PreferenceManager.getDefaultSharedPreferences(context);
        editor = preferences.edit();
    }

    public SharedPreferencesHelper(Context context, String name) {
        super(context);
        preferences = getSharedPreferences(name, Context.MODE_PRIVATE);
        editor = preferences.edit();
    }

    public static SharedPreferencesHelper getPreferencesHelper(Context context) {
        return new SharedPreferencesHelper(context, PREFERENCES);
    }

    public Boolean getBoolean(String key) {
        return preferences.getBoolean(key, DEFAULT_BOOLEAN);
    }

    public Boolean getBoolean(String key, boolean defaultValue) {
        return preferences.getBoolean(key, defaultValue);
    }

    public Boolean getBooleanOrNull(String key) {
        return preferences.contains(key) ? getBoolean(key) : null;
    }

    public boolean putBoolean(String key, boolean value) {
        return editor.putBoolean(key, value).commit();
    }

    public void putBooleanSafely(String key, boolean value) {
        editor.putBoolean(key, value).apply();
    }

    public Float getFloat(String key) {
        return preferences.getFloat(key, DEFAULT_FLOAT);
    }

    public Float getFloat(String key, float defaultValue) {
        return preferences.getFloat(key, defaultValue);
    }

    public Float getFloatOrNull(String key) {
        return preferences.contains(key) ? getFloat(key) : null;
    }

    public boolean putFloat(String key, float value) {
        return editor.putFloat(key, value).commit();
    }

    public void putFloatSafely(String key, float value) {
        editor.putFloat(key, value).apply();
    }

    public Integer getInt(String key) {
        return preferences.getInt(key, DEFAULT_INTEGER);
    }

    public Integer getInt(String key, int defaultValue) {
        return preferences.getInt(key, defaultValue);
    }

    public Integer getIntOrNull(String key) {
        return preferences.contains(key) ? getInt(key) : null;
    }

    public boolean putInt(String key, int value) {
        return editor.putInt(key, value).commit();
    }

    public void putIntSafely(String key, int value) {
        editor.putInt(key, value).apply();
    }

    public Long getLong(String key) {
        return preferences.getLong(key, DEFAULT_LONG);
    }

    public Long getLong(String key, long defaultValue) {
        return preferences.getLong(key, defaultValue);
    }

    public Long getLongOrNull(String key) {
        return preferences.contains(key) ? getLong(key) : null;
    }

    public boolean putLong(String key, long value) {
        return editor.putLong(key, value).commit();
    }

    public void putLongSafely(String key, long value) {
        editor.putLong(key, value).apply();
    }

    public String getString(String key) {
        return preferences.getString(key, DEFAULT_STRING);
    }

    public String getString(String key, String defaultValue) {
        return preferences.getString(key, defaultValue);
    }

    public String getStringOrNull(String key) {
        return preferences.contains(key) ? getString(key) : null;
    }

    public int getStringAsInt(String key) {
        return getStringAsInt(key, DEFAULT_INTEGER);
    }

    public int getStringAsInt(String key, int defaultValue) {
        if (key == null)
            return defaultValue;

        try {
            return Integer.parseInt(getString(key, DEFAULT_STRING));
        } catch (NumberFormatException ignored) {
            return defaultValue;
        }
    }

    public boolean putString(String key, String value) {
        return editor.putString(key, value).commit();
    }

    public void putStringSafely(String key, String value) {
        editor.putString(key, value).apply();
    }

    public Set<String> getStringSet(String key) {
        return preferences.getStringSet(key, DEFAULT_STRING_SET);
    }

    public Set<String> getStringSet(String key, Set<String> defaultValue) {
        return preferences.getStringSet(key, defaultValue);
    }

    public Set<String> getStringSetOrNull(String key) {
        return preferences.contains(key) ? getStringSet(key) : null;
    }

    public boolean putStringSet(String key, Set<String> value) {
        return editor.putStringSet(key, value).commit();
    }

    public void putStringSetSafely(String key, Set<String> value) {
        editor.putStringSet(key, value).apply();
    }

    public Map<String, ?> getAll(String key) {
        return preferences.getAll();
    }

    public boolean contains(String key) {
        return preferences.contains(key);
    }

    public boolean remove(String key) {
        SharedPreferences.Editor editor = preferences.edit();

        if (preferences.contains(key))
            return editor.remove(key).commit();
        return false;
    }

    public boolean clear() {
        SharedPreferences.Editor editor = preferences.edit();
        return editor.clear().commit();
    }

    public void registerOnSharedPreferenceChangeListener(SharedPreferences.OnSharedPreferenceChangeListener listener) {
        preferences.registerOnSharedPreferenceChangeListener(listener);
    }

    public void unregisterOnSharedPreferenceChangeListener(SharedPreferences.OnSharedPreferenceChangeListener listener) {
        preferences.unregisterOnSharedPreferenceChangeListener(listener);
    }
}

File: helper/src/main/java/com/memoeslink/helper/SharedPreferencesHelper.kt
package com.memoeslink.helper

import android.content.Context
import android.content.ContextWrapper
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import androidx.preference.PreferenceManager

open class SharedPreferencesHelper : ContextWrapper {
    private val preferences: SharedPreferences
    private val editor: SharedPreferences.Editor

    constructor(context: Context?) : super(context) {
        preferences = PreferenceManager.getDefaultSharedPreferences(context!!)
        editor = preferences.edit()
    }

    constructor(context: Context?, name: String?) : super(context) {
        preferences = getSharedPreferences(name, MODE_PRIVATE)
        editor = preferences.edit()
    }

    open fun getBoolean(key: String?): Boolean = preferences.getBoolean(key, DEFAULT_BOOLEAN)

    open fun getBoolean(key: String?, defaultValue: Boolean): Boolean =
        preferences.getBoolean(key, defaultValue)

    open fun getBooleanOrNull(key: String?): Boolean? = if (contains(key)) getBoolean(key) else null

    open fun put(key: String?, value: Boolean): Boolean = putBoolean(key, value)

    open fun putBoolean(key: String?, value: Boolean): Boolean =
        editor.putBoolean(key, value).commit()

    open fun save(key: String?, value: Boolean) = saveBoolean(key, value)

    open fun saveBoolean(key: String?, value: Boolean) = editor.putBoolean(key, value).apply()

    open fun getFloat(key: String?): Float = preferences.getFloat(key, DEFAULT_FLOAT)

    open fun getFloat(key: String?, defaultValue: Float): Float =
        preferences.getFloat(key, defaultValue)

    open fun getFloatOrNull(key: String?): Float? = if (contains(key)) getFloat(key) else null

    open fun put(key: String?, value: Float): Boolean = putFloat(key, value)

    open fun putFloat(key: String?, value: Float): Boolean = editor.putFloat(key, value).commit()

    open fun save(key: String?, value: Float) = saveFloat(key, value)

    open fun saveFloat(key: String?, value: Float) = editor.putFloat(key, value).apply()

    open fun getInt(key: String?): Int = preferences.getInt(key, DEFAULT_INTEGER)

    open fun getInt(key: String?, defaultValue: Int): Int = preferences.getInt(key, defaultValue)

    open fun getIntOrNull(key: String?): Int? = if (contains(key)) getInt(key) else null

    open fun put(key: String?, value: Int): Boolean = putInt(key, value)

    open fun putInt(key: String?, value: Int): Boolean = editor.putInt(key, value).commit()

    open fun save(key: String?, value: Int) = saveInt(key, value)

    open fun saveInt(key: String?, value: Int) = editor.putInt(key, value).apply()

    open fun getLong(key: String?): Long = preferences.getLong(key, DEFAULT_LONG)

    open fun getLong(key: String?, defaultValue: Long): Long =
        preferences.getLong(key, defaultValue)

    open fun getLongOrNull(key: String?): Long? = if (contains(key)) getLong(key) else null

    open fun put(key: String?, value: Long): Boolean = putLong(key, value)

    open fun putLong(key: String?, value: Long): Boolean = editor.putLong(key, value).commit()

    open fun save(key: String?, value: Long) = saveLong(key, value)

    open fun saveLong(key: String?, value: Long) = editor.putLong(key, value).apply()

    open fun getString(key: String?): String? = preferences.getString(key, DEFAULT_STRING)

    open fun getString(key: String?, defaultValue: String?): String? =
        preferences.getString(key, defaultValue)

    open fun getStringOrNull(key: String?): String? = if (contains(key)) getString(key) else null

    open fun getStringAsInt(key: String?): Int = getStringAsInt(key, DEFAULT_INTEGER)

    open fun getStringAsInt(key: String?, defaultValue: Int): Int {
        getString(key)?.let { s ->
            try {
                return Integer.parseInt(s)
            } catch (e: NumberFormatException) {
            }
        }
        return defaultValue
    }

    open fun getStringAsIntOrNull(key: String?): Int? =
        if (contains(key)) getStringAsInt(key) else null

    open fun put(key: String?, value: String?): Boolean = putString(key, value)

    open fun putString(key: String?, value: String?): Boolean =
        editor.putString(key, value).commit()

    open fun save(key: String?, value: String?) = saveString(key, value)

    open fun saveString(key: String?, value: String?) = editor.putString(key, value).apply()

    open fun getStringSet(key: String?): Set<String> =
        preferences.getStringSet(key, DEFAULT_STRING_SET) ?: DEFAULT_STRING_SET

    open fun getStringSet(key: String?, defaultValue: Set<String?>?): Set<String> =
        preferences.getStringSet(key, defaultValue) ?: DEFAULT_STRING_SET

    open fun getStringSetOrNull(key: String?): Set<String>? =
        if (contains(key)) getStringSet(key) else null

    open fun put(key: String?, value: Set<String?>?): Boolean = putStringSet(key, value)

    open fun putStringSet(key: String?, value: Set<String?>?): Boolean =
        editor.putStringSet(key, value).commit()

    open fun save(key: String?, value: Set<String?>?) = saveStringSet(key, value)

    open fun saveStringSet(key: String?, value: Set<String?>?) =
        editor.putStringSet(key, value).apply()

    open fun get(key: String?, defaultValue: Any?): Any? {
        return when (defaultValue) {
            is String -> getString(key, defaultValue)
            is Int -> getInt(key, defaultValue)
            is Boolean -> getBoolean(key, defaultValue)
            is Float -> getFloat(key, defaultValue)
            is Long -> getLong(key, defaultValue)
            is Set<*> -> getStringSet(key, defaultValue as Set<String?>)
            else -> null
        }
    }

    open fun getAll(): Map<String, *> = preferences.all

    open fun contains(key: String?): Boolean = preferences.contains(key)

    open fun remove(key: String?): Boolean {
        val editor = preferences.edit()
        return if (contains(key)) editor.remove(key).commit() else false
    }

    open fun clear(): Boolean {
        val editor = preferences.edit()
        return editor.clear().commit()
    }

    open fun registerOnSharedPreferenceChangeListener(listener: OnSharedPreferenceChangeListener?) {
        preferences.registerOnSharedPreferenceChangeListener(listener)
    }

    open fun unregisterOnSharedPreferenceChangeListener(listener: OnSharedPreferenceChangeListener?) {
        preferences.unregisterOnSharedPreferenceChangeListener(listener)
    }

    companion object {
        @JvmStatic val PREFERENCES = "prefs"
        @JvmStatic val DEFAULT_BOOLEAN = false
        @JvmStatic val DEFAULT_FLOAT = 0F
        @JvmStatic val DEFAULT_INTEGER = 0
        @JvmStatic val DEFAULT_LONG = 0L
        @JvmStatic val DEFAULT_STRING = ""
        @JvmStatic val DEFAULT_STRING_SET = emptySet<String>()

        fun getPreferencesHelper(context: Context?): SharedPreferencesHelper {
            return SharedPreferencesHelper(context, PREFERENCES)
        }
    }
}